source/reflect.h2

```diff
@@ -634,6 +719,82 @@ add_virtual_destructor: (inout t: meta::type_declaration) =
     t.add_member( "operator=: (virtual move this) = { }");
 }
 
+//-----------------------------------------------------------------------
+//
+//  foreign_interface_pseudovalue
+//
+//  Internal metafunction to help with public reflection API declaration.
+//
+//    * "foreign interface" because the interface its declaring will be
+//      defined somewhere else outside the metaprogram.
+//
+//    * "pseudovalue" because the resulting type behaves like a value,
+//      you can copy, move it, etc. But its not an actual value type,
+//      as all of its "guts" are behind pointers to the implementation.
+//
+//  The verbose name is on purpose, lets try to give it a better name if it
+//  can be generalized into something useful for all developers.
+foreign_interface_pseudovalue: (inout t : meta::type_declaration) =
```

Comment:
I should have explained this better in the PR summary, but this is not just for an interface, we are doing more than that.

The current API uses value semantics, not pointer/reference semantics. For example, when you call `declaration.as_type()`, you are not getting a reference to a declaration, you are getting a value representation of the type declaration (`type_declaration`, not `type_declaration&` or `type_declaration*`). This presents an issue since in order to have a full value instance like this, you must know the entire definition, which in turn would mean spilling all the implementation guts of the compiler to the users of the reflection API.

Instead, by using this specific metafunction, we avoid the issue entirely by making the definitions "wrappers" that point to the actual implementation, which lives within Cppfront (or anything else that might implement the reflection API). You can't do that with just a regular `@interface`, since you'd need to cast to a pointer, and modify every metafunction to conform.

Reply:
Ok, I understand the problem. So the options are:
 1. Have this metafunction that basically is a value wrapped interface. (Maybe rename it to something like this.)
 2. Change the call syntax of metafunctions to either use pointers or `inout` style. 

My own choice would be 2 with `inout` style. But I can see also the advantages of option 1. Since the compiler services are already provided by pointer maybe we can move everything to pointer semantics.

I just thought about the reason why Herb might have chosen value semantics here. The objects are constructed on the fly. Using pointers or references would not work since the objects are only temporary and would be destroyed. This could only be solved with smartpointers, which we do not want to have for such an interface. @hsutter  is this correct?

If so, your solution should be generalized.